src/notification/index.js

```diff
@@ -71,23 +71,44 @@ export const getAccountFromNotificationData = (
     return null;
   }
 
-  if (urlMatches.length > 1) {
-    // The user has several accounts in the notification's realm.  We should
-    // be able to tell the right one using the notification's `user_id`...
-    // except we don't store user IDs in `accounts`, only emails.  Until we
-    // fix that, just ignore the information.
-    logging.warn('notification realm_uri ambiguous; multiple matches found', {
-      realm_uri,
-      parsed_url: realmUrl,
-      match_count: urlMatches.length,
-      unique_identities_count: new Set(urlMatches.map(matchIndex => identities[matchIndex].email))
-        .size,
-    });
-    // TODO get user_id into accounts data, and use that
-    return null;
+  // TODO(server-2.1): Remove this, because user_id will always be present
+  if (user_id === undefined) {
+    if (urlMatches.length > 1) {
+      logging.warn(
+        'notification realm_uri ambiguous; multiple matches found; user_id missing (old server)',
+        {
+          realm_uri,
+          parsed_url: realmUrl,
+          match_count: urlMatches.length,
+          unique_identities_count: new Set(urlMatches.map(matchIndex => accounts[matchIndex].email))
+            .size,
+        },
+      );
+      return null;
+    } else {
+      return urlMatches[0];
+    }
   }
 
-  return urlMatches[0];
+  // There may be multiple accounts in the notification's realm. Pick one
+  // based on the notification's `user_id`.
+  const userMatch = urlMatches.find(urlMatch => {
+    const { userId } = accounts[urlMatch];
+    return userId !== null && userId === user_id;
+  });
+  if (userMatch == null) {
+    // Maybe we didn't get a userId match because the correct account just
+    // hasn't had its userId recorded on it yet. See jsdoc on the Account
+    // type for when that is.
+    const nullUserIdMatches = urlMatches.filter(urlMatch => accounts[urlMatch].userId === null);
+    if (nullUserIdMatches.length === 1) {
+      return nullUserIdMatches[0];
+    }
+
+    logging.warn('notification account not found with realm_uri and user_id');
+    return null;
```

Comment:
This log message won't always be accurate -- there could have been two accounts that had the right realm_uri and null user_id.

(Well, it could be read as accurate depending on how one interprets "found with realm_uri and user_id". But a reading that makes it accurate would make the same message equally accurate in the case where there's a match we actually use, just above here.)

We've been pretty precise in the different error cases in this function so far, because it's been a fairly common set of errors and it's good to know exactly what conditions are driving that. So let's make the distinction here too between zero matches and multiple matches.